### PR TITLE
Support expression for EPHEMERAL (currently only literal is supported)

### LINKED ISSUE #44010: Ephemeral columns with Map type crash the server
**Describe what's wrong**

Creating a table with a Map-type Ephemeral column will make the server crash on the restart because of wrongly save metadata.

Fiddle to reproduce: https://fiddle.clickhouse.com/81fc3a83-f7ea-43c2-afad-2e2940d04e8c

**Does it reproduce on recent release?**

Tested on version 22.11.1.1286.

**How to reproduce**

* Use version 22.11.x
* Use the fiddle above to reproduce.

**Expected behavior**

When creating a `EPHEMERAL` columns, a default value is assigned.
Example with a Int32:
```sql
CREATE TABLE name (
  example_col Int32 EPHEMERAL
) -- ... engine stuff here ...
```

Metadata will assign a default value:
```sql
CREATE TABLE name (
  example_col Int32 EPHEMERAL 0
) -- ... engine stuff here ...
```

But, with Map-type, no default value gets assigned, only a `()` (empty parenthesis).

**Error message and/or stacktrace**

```
2022.12.07 14:59:29.359153 [ 11464597 ] {} <Error> Application: DB::Exception: Syntax error (in file /Users/vlourme/Desktop/ch/store/75e/75e717de-1bbd-4011-802c-acd04e9a3409/bug_ephemeral.sql): failed at position 118 ('(') (line 4, col 41): (),
    `result` Int32 DEFAULT (data['number']) * 2
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192
. Expected one of: literal, NULL, number, Bool, true, false, string literal, NOT, COMMENT, CODEC, TTL, token, Comma, ClosingRoundBracket: Cannot parse definition from metadata file /Users/vlourme/Desktop/ch/store/75e/75e717de-1bbd-4011-802c-acd04e9a3409/bug_ephemeral.sql
```

**Additional context**

The issue can't be fixed by setting a default expression after `EPHEMERAL`, this, for example will not work:
```sql
CREATE TABLE bug_ephemeral (
    id Int32,
    data Map(String, Int32) EPHEMERAL {},
    result Int DEFAULT data['number'] * 2
)
Engine = MergeTree()
ORDER BY (id);
```

But this, is accepted:
```sql
CREATE TABLE bug_ephemeral (
    id Int32,
    data Int32 EPHEMERAL 12,
    result Int DEFAULT data * 2
)
Engine = MergeTree()
ORDER BY (id);
```

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed bug with non-parsable default value for EPHEMERAL column in table metadata.

closes #44010

### CODE REVIEW

File: src/Parsers/ASTColumnDeclaration.cpp [hunk 1]
```diff
@@ -78,7 +79,12 @@ void ASTColumnDeclaration::formatImpl(const FormatSettings & settings, FormatSta
     if (default_expression)
     {
         settings.ostr << ' ' << (settings.hilite ? hilite_keyword : "") << default_specifier << (settings.hilite ? hilite_none : "");
-        if (default_specifier != "EPHEMERAL" || !default_expression->as<ASTLiteral>()->value.isNull())
+        if (default_specifier != "EPHEMERAL" ||
+            !(
+                    default_expression->as<ASTLiteral>()->value.isNull() ||
+                    (type && default_expression->as<ASTLiteral>()->value == DataTypeFactory::instance().get(type)->getDefault())
+            )
+        )
```
Comment (antonio2368) [inline src/Parsers/ASTColumnDeclaration.cpp:87]: Weird formatting

Comment (antonio2368) [reply src/Parsers/ASTColumnDeclaration.cpp:87]: Also maybe store different parts in a boolean variable, it's really hard to follow what's going on.

**Review by antonio2368 — COMMENTED** [review]
Please add a stateless test with `SHOW CREATE TABLE`.

File: src/Parsers/ASTColumnDeclaration.cpp (hunk 1 above)
Comment (yakov-olkhovskiy) [reply src/Parsers/ASTColumnDeclaration.cpp]: I refactored it - the previous implementation was quite deficient...

File: src/Parsers/ASTColumnDeclaration.cpp [hunk 2]
```diff
@@ -78,7 +79,7 @@ void ASTColumnDeclaration::formatImpl(const FormatSettings & settings, FormatSta
     if (default_expression)
     {
         settings.ostr << ' ' << (settings.hilite ? hilite_keyword : "") << default_specifier << (settings.hilite ? hilite_none : "");
-        if (default_specifier != "EPHEMERAL" || !default_expression->as<ASTLiteral>()->value.isNull())
+        if (!ephemeral_default)
```
Comment (antonio2368) [inline src/Parsers/ASTColumnDeclaration.cpp:82]: I'm a bit confused why we can't have
```suggestion
        if (default_specified != "EPHEMERAL")
```
because this `ephemeral_default` represents a presence of default_expression for `EPHEMERAL` default_specifier

Comment (yakov-olkhovskiy) [reply src/Parsers/ASTColumnDeclaration.cpp:82]: we have special meaning for `EPHEMERAL` different from `DEFAULT` - `EPHEMERAL` signifies that column is ephemeral, and then it may have optional expression - if this expression present - it's a default value, if not - default is a default value of column's type - so, the flag `ephemeral_default` means that this expression is not present (and value is defaulted to type default) - which means we don't want to print it out at all. However if `EPHEMERAL` contains expression, then `ephemeral_default` is false and we want to print this expression. For default_specifier `DEFAULT` expression is always present, `ephemeral_default` is not changed and always false, and we always want to print this expression - so, from `ephemeral_default` we always know if we want to print this expression

**Review by antonio2368 — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
@Felixoid I linked dbms into parsers examples (select_parser and create_parser) - it will inflate the executables size twice - I think we don't actually need examples artefacts to be stored, so probably we need somehow exclude them...
